pkg/binding/crd.go

```diff
@@ -35,19 +38,55 @@ type APIResource struct {
 }
 
 // Handle CRDs should account for CRDs being added or deleted to start/stop new informers as needed
-func (c *Controller) handleCRD(_ runtime.Object) error {
-	toStartList, toStopList, err := c.checkAPIResourcesForUpdates()
-	if err != nil {
-		return err
+func (c *Controller) handleCRD(ctx context.Context, obj runtime.Object) error {
+	logger := klog.FromContext(ctx)
+	uObj := obj.(*unstructured.Unstructured)
+	var crdObj *apiextensionsv1.CustomResourceDefinition
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(uObj.UnstructuredContent(), &crdObj); err != nil {
+		return fmt.Errorf("failed to convert Unstructured to CRD: %w", err)
 	}
 
-	go c.startInformersForNewAPIResources(toStartList)
+	// for each CustomResourceDefinitionVersion, follow a decision tree to tell whether the corresponding gvr should be watched
+	toStartList, toStopList := []APIResource{}, []string{}
+	for _, ver := range crdObj.Spec.Versions {
+		gvr := APIResource{
+			groupVersion: schema.GroupVersion{
+				Group:   crdObj.Spec.Group,
+				Version: ver.Name,
+			},
+			resource: metav1.APIResource{
+				Name: crdObj.Spec.Names.Plural,
+				Kind: crdObj.Spec.Names.Kind,
+			},
+		}
+		key := util.KeyForGroupVersionKind(gvr.groupVersion.Group, gvr.groupVersion.Version, gvr.resource.Kind)
+		if isBeingDeleted(obj) {
+			toStopList = append(toStopList, key)
+			continue
+		}
+		if !c.includedToWatch(gvr) {
+			toStopList = append(toStopList, key)
+			continue
+		}
+		if !ver.Served {
+			toStopList = append(toStopList, key)
+			continue
+		}
+		if crdEstablished(crdObj) {
+			toStartList = append(toStartList, gvr)
+		}
+	}
+
+	if len(toStartList) > 0 {
+		go c.startInformersForNewAPIResources(ctx, toStartList)
+	}
 
 	for _, key := range toStopList {
-		c.logger.Info("API removed, stopping informer.", "key", key)
-		stopper := c.stoppers[key]
-		// close channel
-		close(stopper)
+		logger.Info("API should not be watched, stopping informer.", "key", key)
```

Comment:
This log message is misleading when `ok` in the next statement is `false`.

Reply:
Revised.